README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
humanize
========

[![Travis-CI Build Status](https://travis-ci.org/newtux/humanize.svg?branch=master)](https://travis-ci.org/newtux/humanize) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/humanize)](https://cran.r-project.org/package=humanize)

Humanize is an almost direct port of the python [humanize package](https://github.com/jmoiron/humanize).

The goal of humanize is to provide some utlities in order to turn values (so far times, file sizes, and numbers) into human readable forms.

Installation
------------

You can install the latest CRAN version with:

``` r
install.packages("humanize")
```

You can install humanize from github with:

``` r
# install.packages("devtools")
devtools::install_github("newtux/humanize")
```

Examples
--------

### Times

Convert times:

``` r
library(humanize)

natural_time(Sys.time())
#> [1] "now"
natural_time(Sys.time() - 1)
#> [1] "a second ago"
natural_time(Sys.time() - 100)
#> [1] "a minute ago"
natural_time(Sys.time() - 1000*10)
#> [1] "2 hours ago"
```

Works across days:

``` r
natural_time(Sys.time() - lubridate::ddays(1))
#> [1] "a day ago"
natural_time(Sys.time() - lubridate::ddays(70))
#> [1] "2 months ago"
```

And forward in time:

``` r
natural_time(Sys.time() + lubridate::ddays(1))
#> [1] "23 hours from now"
```

### File Sizes

Convert file sizes:

``` r
natural_size(300)
#> 300 Bytes
natural_size(3000)
#> 3.0 kB
natural_size(3000000)
#> 3.0 MB
natural_size(3000000000)
#> 3.0 GB
natural_size(3000000000000)
#> 3.0 TB
natural_size(10**26 * 30)
#> 3000.0 YB
```

### Numbers

Ordinals:

``` r
count_as_ordinal(1)
#> [1] "1st"
count_as_ordinal(111)
#> [1] "111th"
```

Comma Seperation:

``` r
number_as_comma(1000)
#> [1] "1,000"
number_as_comma(10000)
#> [1] "10,000"
```

Words:

``` r
count_as_word(100)
#> [1] "100"
count_as_word(1000000)
#> [1] "1.0 million"
count_as_word(1200000000)
#> [1] "1.2 billion"
```

AP Format:

``` r
count_as_ap(3)
#> [1] "three"
count_as_ap(20)
#> [1] "20"
```

Todo
----

This is still a very early cut of the package.

-   Better support in times? For diff time?
-   Maybe add times relative to other times?
-   Export helper functions used in tests?


NEWS.md
# humanize 0.2.0

* Ported over `number` functions
* Clean up arguments to be consistent across library (format -> fmt)
* Fix bugs with tests

# humanize 0.1.0

* Ported over `time` functions
* Ported over `filesize` functions
* Added a `NEWS.md` file to track changes to the package.


tests/testthat/test_time.R

#useful time constants
today <- Sys.Date()
tomorrow <- today + 1
yesterday <- today - 1
past_date <- as.Date('2015-06-01')
minute <- 60
hour <- minute*60
day <- hour*24
month <- day*30.5
year <- 365*day

get_seconds_helper <- function(seconds=0, minutes=0, hours=0, days=0) {
  dur <- lubridate::dseconds(seconds) +
    lubridate::dminutes(minutes) +
    lubridate::dhours(hours) +
    lubridate::ddays(days)

  lubridate::time_length(dur, unit = "second")
}

get_duration_helper <- function(seconds=0, minutes=0, hours=0, days=0) {
  dur <- lubridate::dseconds(seconds) +
    lubridate::dminutes(minutes) +
    lubridate::dhours(hours) +
    lubridate::ddays(days)

  dur
}

test_that("seconds_to_natural_delta no months works as expected", {
  test_input = c(
    day*7,
    day*31,
    day*230,
    day*400
  )
  test_output <- test_input %>%
    purrr::map_chr(seconds_to_natural_delta, use_months = FALSE)
  result_list = c(
    '7 days',
    '31 days',
    '230 days',
    '1 year, 35 days'
  )

  purrr::walk2(test_output, result_list, expect_match)
})

test_that("seconds_to_natural_delta with months works as expected", {
  test_input = c(
    0,
    1,
    30,
    get_seconds_helper(minutes = 1, seconds = 30),
    get_seconds_helper(minutes = 2),
    get_seconds_helper(hours = 1, minutes = 30, seconds = 30),
    get_seconds_helper(hours = 23, minutes = 50, seconds = 50),
    get_seconds_helper(days = 1),
    get_seconds_helper(days = 500),
    get_seconds_helper(days = 365*2 + 35),
    get_seconds_helper(seconds = 1),
    get_seconds_helper(seconds = 30),
    get_seconds_helper(minutes = 1, seconds = 30),
    get_seconds_helper(minutes = 2),
    get_seconds_helper(hours = 1, minutes = 30, seconds = 30),
    get_seconds_helper(hours = 23, minutes = 50, seconds = 50),
    get_seconds_helper(days = 1),
    get_seconds_helper(days = 500),
    get_seconds_helper(days = 365*2 + 35),
    get_seconds_helper(days = 10000),
    get_seconds_helper(days = 365 + 35),
    get_seconds_helper(days = 365*2 + 65),
    get_seconds_helper(days = 365 + 4),
    get_seconds_helper(days = 35),
    get_seconds_helper(days = 65),
    get_seconds_helper(days = 9),
    get_seconds_helper(days = 365)
  )
  test_output <- test_input %>%
    purrr::map_chr(seconds_to_natural_delta)
  result_list = c(
    'now',
    'a second',
    '30 seconds',
    'a minute',
    '2 minutes',
    'an hour',
    '23 hours',
    'a day',
    '1 year, 4 months',
    '2 years',
    'a second',
    '30 seconds',
    'a minute',
    '2 minutes',
    'an hour',
    '23 hours',
    'a day',
    '1 year, 4 months',
    '2 years',
    '27 years',
    '1 year, 1 month',
    '2 years',
    '1 year, 4 days',
    'a month',
    '2 months',
    '9 days',
    'a year'
  )
  purrr::walk2(test_output, result_list, expect_match)
})


test_that("natural_time works as expected", {
  right_now <- lubridate::now()
  # need to use list here so we can mix ints and times
  test_list <- list(
    right_now,
    right_now - get_duration_helper(seconds = 1),
    right_now - get_duration_helper(seconds = 30),
    right_now - get_duration_helper(minutes = 1, seconds = 30),
    right_now - get_duration_helper(minutes = 2),
    right_now - get_duration_helper(hours = 1, minutes = 30, seconds = 30),
    right_now - get_duration_helper(hours = 23, minutes = 50, seconds = 50),
    right_now - get_duration_helper(days = 1),
    right_now - get_duration_helper(days = 500),
    right_now - get_duration_helper(days = 365*2 + 35),
    right_now + get_duration_helper(seconds = 1),
    right_now + get_duration_helper(seconds = 30),
    right_now + get_duration_helper(minutes = 1, seconds = 30),
    right_now + get_duration_helper(minutes = 2),
    right_now + get_duration_helper(hours = 1, minutes = 30, seconds = 30),
    right_now + get_duration_helper(hours = 23, minutes = 50, seconds = 50),
    right_now + get_duration_helper(days = 1),
    right_now + get_duration_helper(days = 500),
    right_now + get_duration_helper(days = 365*2 + 35),
    right_now + get_duration_helper(days = 10000),
    right_now - get_duration_helper(days = 365 + 35),
    30,
    right_now - get_duration_helper(days = 365*2 + 65),
    right_now - get_duration_helper(days = 365 + 4)
  )
  # need to fix the current  time
  with_mock(
    now  = function(x) {right_now},
    test_output <- test_list %>%
      purrr::map_chr(natural_time),
    .env = "lubridate"
  )

  result_list <- c(
    'now',
    'a second ago',
    '30 seconds ago',
    'a minute ago',
    '2 minutes ago',
    'an hour ago',
    '23 hours ago',
    'a day ago',
    '1 year, 4 months ago',
    '2 years ago',
    'a second from now',
    '30 seconds from now',
    'a minute from now',
    '2 minutes from now',
    'an hour from now',
    '23 hours from now',
    'a day from now',
    '1 year, 4 months from now',
    '2 years from now',
    '27 years from now',
    '1 year, 1 month ago',
    '30 seconds ago',
    '2 years ago',
    '1 year, 4 days ago'
  )
  purrr::walk2(test_output, result_list, expect_match)
})

test_that("natural_time no months works as expected", {
  right_now <- lubridate::now()
  # need to use list here so we can mix ints and times
  test_list = list(
    right_now,
    right_now - get_duration_helper(seconds = 1),
    right_now - get_duration_helper(seconds = 30),
    right_now - get_duration_helper(minutes = 1, seconds = 30),
    right_now - get_duration_helper(minutes = 2),
    right_now - get_duration_helper(hours = 1, minutes = 30, seconds = 30),
    right_now - get_duration_helper(hours = 23, minutes = 50, seconds = 50),
    right_now - get_duration_helper(days = 1),
    right_now - get_duration_helper(days = 17),
    right_now - get_duration_helper(days = 47),
    right_now - get_duration_helper(days = 500),
    right_now - get_duration_helper(days = 365*2 + 35),
    right_now + get_duration_helper(seconds = 1),
    right_now + get_duration_helper(seconds = 30),
    right_now + get_duration_helper(minutes = 1, seconds = 30),
    right_now + get_duration_helper(minutes = 2),
    right_now + get_duration_helper(hours = 1, minutes = 30, seconds = 30),
    right_now + get_duration_helper(hours = 23, minutes = 50, seconds = 50),
    right_now + get_duration_helper(days = 1),
    right_now + get_duration_helper(days = 500),
    right_now + get_duration_helper(days = 365*2 + 35),
    right_now + get_duration_helper(days = 10000),
    right_now - get_duration_helper(days = 365 + 35),
    30,
    right_now - get_duration_helper(days = 365*2 + 65),
    right_now - get_duration_helper(days = 365 + 4)
  )

  # need to fix the current  time
  with_mock(
    now  = function(x) {right_now},
    test_output <- test_list %>%
      purrr::map_chr(natural_time, use_months = FALSE),
    .env = "lubridate"
  )

  result_list = c(
    'now',
    'a second ago',
    '30 seconds ago',
    'a minute ago',
    '2 minutes ago',
    'an hour ago',
    '23 hours ago',
    'a day ago',
    '17 days ago',
    '47 days ago',
    '1 year, 135 days ago',
    '2 years ago',
    'a second from now',
    '30 seconds from now',
    'a minute from now',
    '2 minutes from now',
    'an hour from now',
    '23 hours from now',
    'a day from now',
    '1 year, 135 days from now',
    '2 years from now',
    '27 years from now',
    '1 year, 35 days ago',
    '30 seconds ago',
    '2 years ago',
    '1 year, 4 days ago'
  )
  purrr::walk2(test_output, result_list, expect_match)
})


test_that("natural_day works as expected", {


  expect_match(natural_day(today), 'today')
  expect_match(natural_day(tomorrow), 'tomorrow')
  expect_match(natural_day(yesterday), 'yesterday')
  expect_match(natural_day(past_date), "Jun 01")
  # TODO Error tests

})


test_that("natural_date works as expected", {
  past_date <- as.Date('2015-06-01')

  expect_match(natural_date(today), 'today')
  expect_match(natural_date(tomorrow), 'tomorrow')
  expect_match(natural_date(yesterday), 'yesterday')
  expect_match(natural_date(past_date), "Jun 01 2015")
  # TODO Error tests

})


tests/testthat/test_number.R
context("test_number.R")

test_that("number_as_comma works as expected", {
  test_input <- c(100, 1000, 10123, 10311, 1000000, 1234567.25, 100,
                1000, 10123, 10311, 1000000, 1234567.1234567)
  test_output <- test_input %>%
    purrr::map_chr(number_as_comma)
  results <- c('100', '1,000', '10,123', '10,311', '1,000,000',
                '1,234,567.25', '100', '1,000', '10,123', '10,311', '1,000,000',
                '1,234,567.1234567')
  purrr::walk2(test_output, results, expect_match)
})


test_that("count_as_ap works as expected", {
  test_input <- c(1, 2, 4, 5, 9, 10, 1000000000)
  test_output <- test_input %>%
    purrr::map_chr(count_as_ap)
  results <- c("one", "two", "four", "five", "nine", "10", "1000000000")

  purrr::walk2(test_output, results, expect_match)

  expect_error(count_as_ap(-1))
  expect_error(count_as_ap("test"))

})

test_that("count_as_word works as expected", {
  test_input <- c(100, 1000000, 1200000, 1290000, 1000000000,
                  2000000000)
  test_output <- test_input %>%
    purrr::map_chr(count_as_word)
  results <- c('100', '1.0 million', '1.2 million', '1.3 million',
              '1.0 billion', '2.0 billion')

  purrr::walk2(test_output, results, expect_match)
})


tests/testthat/test_filesize.R
test_that("natural_size decimal works as expected", {
  test_input <- c(300, 3000, 3000000, 3000000000, 3000000000000,10**26 * 30)

  test_output <- test_input %>%
    purrr::map_chr(natural_size)
  results <- c('300 Bytes', '3.0 kB', '3.0 MB', '3.0 GB', '3.0 TB','3000.0 YB')
  purrr::walk2(test_output, results, expect_match)
})

test_that("natural_size binary works as expected", {
  test_input <- c(300, 3000, 3000000, 10**26 * 30)
  test_output <- test_input %>%
    purrr::map_chr(natural_size, suffix_type = 'binary')
  results <- c('300 Bytes', '2.9 KiB', '2.9 MiB',  '2481.5 YiB')
  purrr::walk2(test_output, results, expect_match)
})

test_that("natural_size gnu works as expected", {
  test_input <- c(300, 3000, 3000000, 1024, 10**26 * 30)
  test_output <- test_input %>%
    purrr::map_chr(natural_size, suffix_type =  'gnu')
  results <- c('300B', '2.9K', '2.9M', '1.0K', '2481.5Y')

  purrr::walk2(test_output, results, expect_match)
})

test_that("natural_size format passing works as expected", {
  expect_match(natural_size(3141592, 'decimal', '%.2f'), '3.14 MB')
  expect_match(natural_size(3000, 'gnu', '%.3f'), '2.930K')
  expect_match(natural_size(3000000000, 'gnu', '%.0f'), '3G')
  expect_match(natural_size(10**26 * 30, 'binary', '%.3f'), '2481.542 YiB')
})


R/number.R

ordinals <- list(
  "0" = "th",
  "1" = "st",
  "2" = "nd",
  "3" = "rd",
  "4" = "th",
  "5" = "th",
  "6" = "th",
  "7" = "th",
  "8" = "th",
  "9" = "th"
)

#' Transform a count to an ordinal string
#'
#' @param value A single positive integer
#'
#' @return A string with the ordinal reprisentation of a number
#' @export
#'
#' @examples
#' count_as_ordinal(1)
#' count_as_ordinal(111)
count_as_ordinal <- function(value) {
  assert_that(is.count(value))

  if (value %% 100 %in% c(11, 12, 13)) {
    return(paste0(value, ordinals[["0"]]))
  }
  paste0(value, ordinals[[as.character(value %% 10)]])
}

#' Convert an number to a string with comma seperation
#'
#' Just a wrapper around `format` with defaults for full digits
#'
#' @param value A numeric
#'
#' @return A string with comma separation every three digits
#' @export
#'
#' @examples
#' number_as_comma(1000)
#' number_as_comma(10000)
number_as_comma <- function(value) {
  assert_that(is.numeric(value))
  format(value, big.mark = ",", scientific = FALSE, digits = 22)
}

human_powers <- 10^c(6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 100)
names(human_powers) <- c(
  'million','billion','trillion','quadrillion',
  'quintillion','sextillion','septillion',
  'octillion','nonillion','decillion','googol'
)

#' Convert Large Counts into Friendly Text
#'
#' Note - currently limited to `.Machine$integer.max`.
#'
#' @param value A single positive integer
#' @param fmt Extra number formatting supplied to sprintf
#'
#' @return Returns a string with the power of a number replaced by the appropriate word.
#' @export
#'
#' @examples
#' count_as_word(100)
#' count_as_word(1000000)
#' count_as_word(1200000000)
count_as_word <- function(value, fmt='%.1f') {
  assert_that(is.count(value))

  if (value < human_powers[1]) {
    return(format(value, scientific = FALSE))
  }
  # Drop the first value, but shift our index
  for (i in seq_along(human_powers[-1]) + 1) {
    if (value < human_powers[[i]]) {
      chopped <- value / human_powers[[i - 1]]
      human_name <- names(human_powers[i - 1])
      return(paste(sprintf(fmt, chopped), human_name))
    }
  }
  # gigantic values
  # R does not seem to have percision here. TODO?
  return(format(value, scientific = FALSE))
}


ap_numbers <- c("one", "two", "three", "four", "five", "six", "seven", "eight", "nine")


#' Convert to AP Number
#'
#' @param value A single positive integer
#'
#' @return For numbers 1-9, returns the number spelled out. Otherwise, returns
#'   the number as a string.
#' @export
#'
#' @examples
#' count_as_ap(3)
#' count_as_ap(20)
count_as_ap <- function(value) {
  assert_that(is.count(value))
  if (value >= 10) {
    return(format(value, scientific = FALSE))
  }

  ap_numbers[value]
}

fractional <- function(value) {

}


R/filesize.R


suffixes <- list(
  'decimal' =  c('kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'),
  'binary' = c('KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'),
  'gnu' = c("K","M","G","T","P","E","Z","Y")
)

#' Convert bytes to a more natural representation
#'
#' @param bytes Number of bytes
#' @param suffix_type One of 'decimal', 'binary', 'gnu'
#' @param fmt Extra number formatting
#'
#' @export
#'
#' @examples
#' natural_size(3000)
natural_size <- function(bytes, suffix_type="decimal", fmt='%.1f') {
  stopifnot(suffix_type %in% names(suffixes))
  # How much value check do I need for bytes?
  suffix <- suffixes[[suffix_type]]
  gnu <- suffix_type == "gnu"

  base <- ifelse(suffix_type %in% c('gnu', 'binary'), 1024, 1000)

  if (bytes == 1 & !gnu) {
    return("1 Byte")
  } else if (bytes < base & !gnu) {
    return(glue::glue("{bytes} Bytes"))
  } else if (bytes < base & gnu) {
    return(glue::glue("{bytes}B"))
  }

  for (i in seq_along(suffix)) {
    unit <- base ^ (i + 1)
    if (bytes < unit) {
      out_val <- sprintf(fmt,(base * bytes / unit))
      if (gnu) {
        return(glue::glue("{out_val}{suffix[[i]]}"))
      } else {
        return(glue::glue("{out_val} {suffix[[i]]}"))
      }
    }
  }

  out_val <- sprintf(fmt,(base * bytes / unit))
  if (gnu) {
    return(glue::glue("{out_val}{suffix[[length(suffix)]]}"))
  }
  return(glue::glue("{out_val} {suffix[[length(suffix)]]}"))
}


R/time.R

#' @importFrom lubridate now
lubridate::now

#' @importFrom lubridate today
lubridate::today

#' Takes in a number of seconds and computes a "human" delta
#'
#' @param seconds A positive number of seconds
#' @param use_months Boolean whether we should (imprecisely) use months as a unit
#'
#' @seealso `natural_time`
#' @export
seconds_to_natural_delta <- function(seconds, use_months=TRUE) {
  seconds_in_day <- 60*60*24

  seconds <- abs(seconds)
  days <- trunc(seconds/seconds_in_day)
  years <- trunc(days/365)
  days <- days %% ifelse(years == 0, 365, 365*years)
  months <- trunc(days / 30.5) #interpolate
  # with much thanks to humanize (python)
  if (years == 0) {
    if (days < 1) {
      if (seconds == 0) {
        return("now")
      } else if (seconds == 1) {
        return("a second")
      } else if (seconds < 60) {
        return(glue::glue("{seconds} seconds"))
      } else if (60 <= seconds & seconds < 120) {
        return("a minute")
      } else if ( 120 <= seconds & seconds < 3600) {
        minutes = trunc(seconds / 60)
        return(glue::glue("{minutes} minutes"))
      } else if ( 3600 <= seconds & seconds < (3600 * 2)) {
        return("an hour")
      } else if ( 3600 < seconds) {
        hours = trunc(seconds / 3600)
        return(glue::glue("{hours} hours"))
      }
    } else {
      if (days == 1) {
        return("a day")
      }
      if (!use_months) {
        return(glue::glue("{days} days"))
      } else {
        if (months == 0) {
          return(glue::glue("{days} days"))
        } else if (months == 1) {
          return("a month")
        } else {
          return(glue::glue("{months} months"))
        }
      }
    }
  } else if (years == 1) {
    if (months == 0) {
      if (days == 0) {
        return("a year")
      } else if (days == 1) {
        return(glue::glue("1 year, 1 day"))
      } else {
        return(glue::glue("1 year, {days} days"))
      }
    } else if (use_months) {
      if (months == 1) {
        return("1 year, 1 month")
      } else {
        return(glue::glue("1 year, {months} months"))
      }
    } else {
      return(glue::glue("1 year, {days} days"))
    }
  } else {
    return(glue::glue("{years} years"))
  }
}



#' Convert times to natural values relative to now.
#'
#' Given a datetime or a number of seconds, return a natural representation of
#' that resolution that makes sense. Ago/From now determined by positive or
#' negative values.
#'
#' @param value a datetime or a number of seconds
#' @param use_months Boolean whether we should (imprecisely) use months as a
#'   unit
#'
#' @export
#' @examples
#' natural_time(Sys.time()-1)
#' natural_time(Sys.time()-100)
natural_time <- function(value, use_months=TRUE) {
  if (assertthat::is.number(value)) {
    interval_seconds <- value
  } else {
    time_diff <- lubridate::interval(value, now())
    interval_seconds <- lubridate::time_length(time_diff, "seconds")
  }
  if (interval_seconds >= 0) {
    future <- FALSE
  } else {
    future <- TRUE
  }
  interval_seconds <- abs(interval_seconds)
  natural_delta <- seconds_to_natural_delta(trunc(interval_seconds), use_months)
  if (natural_delta == "now") {
    return(natural_delta)
  }

  rel_str <- ifelse(future, "from now", "ago")
  paste(natural_delta, rel_str)
}



#' Natural Day
#'
#' For date values that are tomorrow, today or yesterday compared to present day
#' returns representing string. Otherwise, returns a string formatted according
#' to `fmt`
#'
#' @param value A date value
#' @param fmt Optional formatting string for dates not yesterday, today, tomorrow
#'
#' @return A nicely formatted date
#' @export
#'
#' @examples
#' natural_day(Sys.Date())
#' natural_day(Sys.Date()-10)
natural_day <- function(value, fmt='%b %d') {
  assert_that(is.date(value))
  delta <- value - today()

  if (delta == 0) {
    out <- 'today'
  } else if (delta == 1) {
    out <- 'tomorrow'
  } else if (delta == -1) {
    out <- 'yesterday'
  } else {
    out <- format(value, fmt)
  }

  out
}

#' Natural Date
#'
#' Like natural day, but will append a year for dates that are a year or
#' more in the past or future
#'
#' @param value A Date value
#'
#' @export
#' @seealso `natural_day`
#' @examples
#' natural_date(Sys.Date())
#' natural_date(Sys.Date()-10)
natural_date <- function(value) {
  assert_that(is.date(value))
  delta <- abs(value - today())
  if (delta >= 365) {
    return(natural_day(value, '%b %d %Y'))
  } else {
    return(natural_day(value))
  }
}
